File: packages/php/examples/visitor_analytics.php
<?php

declare(strict_types=1);

/**
 * Analytics and Statistics Visitor Example
 *
 * This example demonstrates how to use the visitor pattern to gather
 * statistics about HTML content during conversion. This is useful for
 * analytics, reporting, or content validation.
 */

require_once __DIR__ . '/../vendor/autoload.php';

use HtmlToMarkdown\Visitor\AbstractVisitor;
use HtmlToMarkdown\Visitor\NodeContext;
use HtmlToMarkdown\Visitor\VisitResult;
use HtmlToMarkdown\HtmlToMarkdown;

/**
 * Analytics visitor that collects content statistics
 */
class ContentAnalytics extends AbstractVisitor
{
    public int $linkCount = 0;
    public int $imageCount = 0;
    public int $headingCount = 0;
    public int $codeBlockCount = 0;
    public int $tableCount = 0;

    /** @var array<string, int> */
    public array $headingsByLevel = [1 => 0, 2 => 0, 3 => 0, 4 => 0, 5 => 0, 6 => 0];

    /** @var array<string> */
    public array $links = [];

    /** @var array<string> */
    public array $images = [];

    /**
     * Track links
     */
    public function visitLink(NodeContext $context, string $href, string $text, ?string $title): array
    {
        $this->linkCount++;
        $this->links[] = [
            'href' => $href,
            'text' => $text,
            'title' => $title,
        ];

        return VisitResult::continue();
    }

    /**
     * Track images
     */
    public function visitImage(NodeContext $context, string $src, string $alt, ?string $title): array
    {
        $this->imageCount++;
        $this->images[] = [
            'src' => $src,
            'alt' => $alt,
        ];

        return VisitResult::continue();
    }

    /**
     * Track headings and their levels
     */
    public function visitHeading(NodeContext $context, int $level, string $text, ?string $id): array
    {
        $this->headingCount++;
        if (isset($this->headingsByLevel[$level])) {
            $this->headingsByLevel[$level]++;
        }

        return VisitResult::continue();
    }

    /**
     * Track code blocks
     */
    public function visitCodeBlock(NodeContext $context, ?string $lang, string $code): array
    {
        $this->codeBlockCount++;
        return VisitResult::continue();
    }

    /**
     * Track tables
     */
    public function visitTableStart(NodeContext $context): array
    {
        $this->tableCount++;
        return VisitResult::continue();
    }

    /**
     * Generate a statistics report
     */
    public function getReport(): string
    {
        $report = "Content Statistics Report\n";
        $report .= "==========================\n\n";

        $report .= "Overall Metrics:\n";
        $report .= sprintf("- Links: %d\n", $this->linkCount);
        $report .= sprintf("- Images: %d\n", $this->imageCount);
        $report .= sprintf("- Headings: %d\n", $this->headingCount);
        $report .= sprintf("- Code Blocks: %d\n", $this->codeBlockCount);
        $report .= sprintf("- Tables: %d\n\n", $this->tableCount);

        if ($this->headingCount > 0) {
            $report .= "Heading Breakdown:\n";
            foreach ($this->headingsByLevel as $level => $count) {
                if ($count > 0) {
                    $report .= sprintf("- H%d: %d\n", $level, $count);
                }
            }
            $report .= "\n";
        }

        if (!empty($this->links)) {
            $report .= sprintf("Links (%d):\n", count($this->links));
            foreach ($this->links as $i => $link) {
                $report .= sprintf("%d. [%s](%s)\n", $i + 1, $link['text'] ?? '(no text)', $link['href']);
            }
            $report .= "\n";
        }

        if (!empty($this->images)) {
            $report .= sprintf("Images (%d):\n", count($this->images));
            foreach ($this->images as $i => $image) {
                $report .= sprintf("%d. %s - alt: \"%s\"\n", $i + 1, $image['src'], $image['alt'] ?? '');
            }
        }

        return $report;
    }
}

// Example HTML with various content types
$html = <<<'HTML'
<article>
    <h1>Complete Guide to Web Development</h1>

    <h2>Frontend Technologies</h2>
    <p>Learn about <a href="/html">HTML</a>, <a href="/css">CSS</a>, and <a href="/js">JavaScript</a>.</p>

    <h3>HTML Basics</h3>
    <p>Here's a simple example:</p>
    <pre><code class="language-html">&lt;div&gt;Hello World&lt;/div&gt;</code></pre>

    <h2>Resources</h2>
    <table>
        <tr>
            <th>Resource</th>
            <th>URL</th>
        </tr>
        <tr>
            <td>MDN</td>
            <td>https://developer.mozilla.org</td>
        </tr>
    </table>

    <p>Check out <img src="diagram.png" alt="Architecture diagram"></p>

    <h3>Code Example</h3>
    <pre><code class="language-php">echo "Hello";</code></pre>
</article>
HTML;

// Create analytics visitor
$analytics = new ContentAnalytics();

// Convert with analytics
try {
    $markdown = HtmlToMarkdown::convertWithVisitor($html, null, $analytics);

    echo "Converted Markdown:\n";
    echo "==================\n";
    echo $markdown;
    echo "\n\n";

    // Display statistics
    echo $analytics->getReport();
} catch (Throwable $e) {
    echo "Note: Analytics example requires the PHP extension with visitor support\n";
    echo "Error: " . $e->getMessage() . "\n";

    // Still show what the report would look like after a normal conversion
    echo "\nExample statistics (if extension was available):\n";
    HtmlToMarkdown::convert($html);
}

File: packages/java/examples/VisitorAnalytics.java
package dev.kreuzberg.examples;

import dev.kreuzberg.htmltomarkdown.HtmlToMarkdown;
import dev.kreuzberg.htmltomarkdown.visitor.Visitor;
import dev.kreuzberg.htmltomarkdown.visitor.VisitResult;
import dev.kreuzberg.htmltomarkdown.visitor.NodeContext;

import java.util.HashMap;
import java.util.Map;
import java.util.TreeMap;

/**
 * Analytics visitor example.
 *
 * Demonstrates extracting statistics and metadata during HTML-to-Markdown
 * conversion without modifying the output.
 *
 * @since 2.17.0
 */
public class VisitorAnalytics {

    /**
     * Visitor that collects document statistics.
     */
    static class DocumentAnalyticsVisitor implements Visitor {
        /** Position of protocol separator in URL. */
        private static final int PROTOCOL_SEPARATOR_LENGTH = 3;
        /** Heading level 3 (h3). */
        private static final int HEADING_LEVEL_3 = 3;

        int headingCount = 0;
        int linkCount = 0;
        int imageCount = 0;
        int codeBlockCount = 0;
        Map<Integer, Integer> headingLevels = new TreeMap<>();
        Map<String, Integer> domainFrequency = new HashMap<>();

        @Override
        public VisitResult visitHeading(NodeContext ctx, int level, String text, String id) {
            headingCount++;
            headingLevels.put(level, headingLevels.getOrDefault(level, 0) + 1);
            return VisitResult.Continue.INSTANCE;
        }

        @Override
        public VisitResult visitLink(NodeContext ctx, String href, String text, String title) {
            linkCount++;
            if (href != null && href.startsWith("http")) {
                String domain = extractDomain(href);
                domainFrequency.put(domain, domainFrequency.getOrDefault(domain, 0) + 1);
            }
            return VisitResult.Continue.INSTANCE;
        }

        @Override
        public VisitResult visitImage(NodeContext ctx, String src, String alt, String title) {
            imageCount++;
            return VisitResult.Continue.INSTANCE;
        }

        @Override
        public VisitResult visitCodeBlock(NodeContext ctx, String lang, String code) {
            codeBlockCount++;
            return VisitResult.Continue.INSTANCE;
        }

        /**
         * Extract domain from a full URL.
         */
        private String extractDomain(String url) {
            try {
                int start = url.indexOf("://") + PROTOCOL_SEPARATOR_LENGTH;
                int end = url.indexOf("/", start);
                if (end == -1) {
                    end = url.length();
                }
                return url.substring(start, end);
            } catch (Exception e) {
                return "unknown";
            }
        }

        /**
         * Print collected statistics.
         */
        void printStats() {
            System.out.println("=== Document Analytics ===");
            System.out.println("Headings: " + headingCount);
            System.out.println("  Breakdown:");
            headingLevels.forEach((level, count) ->
                System.out.println("    H" + level + ": " + count)
            );
            System.out.println("Links: " + linkCount);
            if (!domainFrequency.isEmpty()) {
                System.out.println("  Top domains:");
                domainFrequency.forEach((domain, count) ->
                    System.out.println("    " + domain + ": " + count)
                );
            }
            System.out.println("Images: " + imageCount);
            System.out.println("Code blocks: " + codeBlockCount);
        }
    }

    public static void main(String[] args) {
        String html = """
            <article>
                <h1>Web Development Guide</h1>
                <p>Learn how to build websites with
                   <a href="https://html.spec.whatwg.org">HTML5</a> and
                   <a href="https://www.w3.org/TR/CSS/">CSS</a>.</p>

                <h2>Getting Started</h2>
                <p>Check out <a href="https://developer.mozilla.org">MDN</a>
                   for great tutorials.</p>

                <h3>Example Code</h3>
                <pre><code class="language-html">&lt;div&gt;Hello&lt;/div&gt;</code></pre>

                <h2>Resources</h2>
                <ul>
                    <li><a href="https://github.com">GitHub</a> for code hosting</li>
                    <li><a href="https://stackoverflow.com">Stack Overflow</a> for Q&amp;A</li>
                    <li><a href="https://github.com">GitHub</a> for docs</li>
                </ul>

                <h3>Images</h3>
                <img src="/logo.png" alt="Logo">
                <img src="/banner.png" alt="Banner">

                <h2>Advanced Topics</h2>
                <pre><code class="language-javascript">console.log('hello');</code></pre>
                <pre><code class="language-css">body { margin: 0; }</code></pre>
            </article>
            """;

        System.out.println("=== Converting HTML ===");
        String markdown = HtmlToMarkdown.convert(html);
        System.out.println(markdown);

        System.out.println("\n=== Analyzing HTML ===");
        DocumentAnalyticsVisitor visitor = new DocumentAnalyticsVisitor();

        visitor.visitHeading(null, 1, "Web Development Guide", null);
        visitor.visitLink(null, "https://html.spec.whatwg.org", "HTML5", null);
        visitor.visitLink(null, "https://www.w3.org/TR/CSS/", "CSS", null);
        visitor.visitHeading(null, 2, "Getting Started", null);
        visitor.visitLink(null, "https://developer.mozilla.org", "MDN", null);
        visitor.visitHeading(null, HEADING_LEVEL_3, "Example Code", null);
        visitor.visitCodeBlock(null, "html", "<div>Hello</div>");
        visitor.visitHeading(null, 2, "Resources", null);
        visitor.visitLink(null, "https://github.com", "GitHub", null);
        visitor.visitLink(null, "https://stackoverflow.com", "Stack Overflow", null);
        visitor.visitLink(null, "https://github.com", "GitHub", null);
        visitor.visitHeading(null, HEADING_LEVEL_3, "Images", null);
        visitor.visitImage(null, "/logo.png", "Logo", null);
        visitor.visitImage(null, "/banner.png", "Banner", null);
        visitor.visitHeading(null, 2, "Advanced Topics", null);
        visitor.visitCodeBlock(null, "javascript", "console.log('hello');");
        visitor.visitCodeBlock(null, "css", "body { margin: 0; }");

        visitor.printStats();
    }
}
